### ISSUE TITLE
Use `tty-link` to turn URLs into terminal hyperlinks

### ISSUE THREAD

**Issue by piotrmurach** [mattbrictson/bundle_update_interactive#57]
Hi Matt 👋 

This is a great extension to the bundler!

As I was playing with it I thought it could use the newly released [tty-link](https://github.com/piotrmurach/tty-link).

For example, the `url` column could display terminal hyperlinks. The `from` and `to` could also be linked to [diffend.io](https://my.diffend.io/gems) to show the changes for review.

The `tty-link` on supported terminals renders a hyperlink, otherwise it falls back on a plain-text version.

To make the `url` column support terminal hyperlinks, I extended the namespace with `tty-link`:

```ruby
require "tty-link"
module BundleUpdateInteractive
  class << self
    attr_accessor :tty_link
    ...
  end
  self.tty_link = TTY::Link.new(plain: ":url")
end
```

And then modified the `lib/bundle_update_interactive/cli/row.rb`:

```ruby
class BundleUpdateInteractive::CLI
  class Row < SimpleDelegator
    def initialize(outdated_gem)
     ...
      @tty_link = BundleUpdateInteractive.tty_link
    end
    def formatted_changelog_uri
      pastel.blue(tty_link.link_to(changelog_uri))
    end
    private
    attr_reader :tty_link
```

Running it on the repository itself and supported terminal:

<img width="1011" alt="bundle_update_interactive_with_hyperlinks" src="https://github.com/user-attachments/assets/24c1a734-c10d-49e6-9604-1307868104b2">

Further modifications could, for example, shorten the URL on supported terminals to a name only:

<img width="588" alt="bundle_update_interactive_with_short_hyperlinks" src="https://github.com/user-attachments/assets/dd685c43-5f16-46a0-a230-a84e793fca19">

I didn't want to submit PR if this isn't useful or doesn't align with how you want the tool to work. Please let me know what you think.

**Comment by mattbrictson** [comment]
@piotrmurach first of all, thanks for all your work on pastel the tty-* gems. I'm obviously a heavy user of them on this project!

Regarding hyperlinks, I don't actually use a terminal that supports them, so I'm not familiar with the common UI patterns and what users are looking for in terms of behavior. I'll need to defer to your best judgement on this.

My intuition tells me we should start small, and simply turn the existing changelog URLs into hyperlinks, like you showed in your first example. However, I believe most terminals already support clicking on URLs without any special hyperlink markup (on macOS Terminal, I can use command+click), so this extra code and dependency would not provide much additional benefit, if that's the only thing we use it for.

I do like the idea of using hyperlinks to make shorter text clickable without displaying the entire URL, like in your second example. If the URL ends with something like `CHANGELOG.md`, or `NEWS.md`, that seems nice and self-explanatory. And it makes the output much narrower and less likely to wrap, which is a great improvement. That seems worth it.

However if the changelog URL points to a GitHub release like https://github.com/mattbrictson/bundle_update_interactive/releases/tag/v0.9.0, shortening the link in the same way would result in `v0.9.0`, which is now a bit ambiguous. Looking at the shortened version, it is no longer immediately clear that it is link to a changelog.

Also, sometimes `update-interactive` can't find a changelog, and in that case the URL points to e.g. the gem home page. We'd have to consider that case as well.

https://github.com/mattbrictson/bundle_update_interactive/blob/a5e0d2b28d919642b2b988353d0ab42dbc68b2ed/lib/bundle_update_interactive/outdated_gem.rb#L39-L49

What do you think? I wonder, do you know of examples of other tools that are doing something similar that we could use as guidance for UI?

What if we show the full URL as the hyperlink (again, like your first example), but then automatically truncate the URL text based on screen width to prevent wrapping? That way we get the benefit of avoiding wrapping, while still showing part or all of the URL when possible to avoid ambiguity. And even if truncated, clicking it would still visit the original full URL. On the other hand, maybe that's too complicated for a small benefit.

Anyway, to summarize: I'm open to a PR for this, but I would like to make sure we get the UI right.

**Comment by piotrmurach** [comment]
@mattbrictson Thanks. ❤️ 

> I believe most terminals already support clicking on URLs without any special hyperlink markup (on macOS Terminal, I can use command+click)

Sure, most terminals can parse URLs and allow you to open them with various click and key combinations. If you feel that's how you want the tool to work then indeed there is no use for `tty-link`. For context, I'll use `tty-link` in the [tty-markdown](https://github.com/piotrmurach/tty-markdown) project. Terminal hyperlinks have advantages over plain text. They include 'natively' highlighted links with custom text, hover effect and support for attributes. I also like that hyperlinks bring a more universal experience well known from the web.

> However if the changelog URL points to a GitHub release like https://github.com/mattbrictson/bundle_update_interactive/releases/tag/v0.9.0, shortening the link in the same way would result in v0.9.0, which is now a bit ambiguous. Looking at the shortened version, it is no longer immediately clear that it is link to a changelog.

I've done only a quick implementation to see how things could work. I could imagine logic that generates link names based on context rather than picking the last part.

> What if we show the full URL as the hyperlink (again, like your first example), but then automatically truncate the URL text based on screen width to prevent wrapping? That way we get the benefit of avoiding wrapping, while still showing part or all of the URL when possible to avoid ambiguity. And even if truncated, clicking it would still visit the original full URL. On the other hand, maybe that's too complicated for a small benefit.

I suggested `tty-link` as an enhancement on top of the current behaviour. I believe truncation is a different problem to solve. Even with the use of terminal hyperlinks, you are going to have the issue on unsupported terminals like Mac default terminal.

> Anyway, to summarize: I'm open to a PR, for this, but I would like to make sure we get the UI right.

I expected this to be a more nuanced or even impossible idea from the start hence this issue rather than PR. I had a quick play and thought `tty-link` could help. Please don't feel any pressure that you need to act on this information. Equally, I appreciate your prompt and detailed reply. I will surely pay attention to your gem and see how the UI develops. But for now, I'll close this and remove the 'burden' off your shoulders of having to think about something you didn't sign up for. All the best open-source creator!
